src/main/java/io/numaproj/numaflow/sink/SinkService.java

```diff
@@ -5,59 +5,107 @@
+                Udsink.Datum handlerDatum = Udsink.Datum.newBuilder()
+                        .setId(d.getId())
+                        .setValue(d.getValue())
+                        .setEventTime(d.getEventTime())
+                        .setWatermark(d.getWatermark())
+                        .build();
 
-        // create response
-        var responseListBuilder = Udsink.ResponseList.newBuilder();
-        for (Response response : responses) {
-            Udsink.Response r = Udsink.Response.newBuilder()
-                    .setId(response.getId())
-                    .setSuccess(response.getSuccess())
-                    .setErrMsg(response.getErr())
-                    .build();
+                try {
+                    sinkDatumStream.WriteMessage(handlerDatum);
+                } catch (InterruptedException e) {
+                    onError(e);
+                }
+            }
 
-            responseListBuilder.addResponses(r);
-        }
+            @Override
+            public void onError(Throwable throwable) {
+                logger.log(Level.WARNING, "Encountered error in sinkFn", throwable);
+                responseObserver.onError(throwable);
+            }
 
-        // set response
-        responseObserver.onNext(responseListBuilder.build());
-        responseObserver.onCompleted();
+            @Override
+            public void onCompleted() {
+                Udsink.ResponseList response = Udsink.ResponseList
+                        .newBuilder()
+                        .getDefaultInstanceForType();
+                try {
+                    sinkDatumStream.WriteMessage(SinkDatumStreamImpl.DONE);
+                    // wait until the sink handler returns, result.get() is a blocking call
+                    List<Response> responses = result.get();
+                    // construct responseList from responses
+                    response = buildResponseList(responses);
+
+                } catch (InterruptedException | ExecutionException e) {
+                    onError(e);
+                }
+                responseObserver.onNext(response);
+                responseObserver.onCompleted();
+            }
+        };
+    }
+
+    // shuts down the executor service which is used for reduce
+    public void shutDown() {
+        this.sinkTaskExecutor.shutdown();
+        try {
+            if (!sinkTaskExecutor.awaitTermination(SHUTDOWN_TIME, TimeUnit.SECONDS)) {
+                System.err.println("Sink executor did not terminate in the specified time.");
+                List<Runnable> droppedTasks = sinkTaskExecutor.shutdownNow();
+                System.err.println("Sink executor was abruptly shut down. " + droppedTasks.size()
+                        + " tasks will not be executed.");
+            } else {
+                System.err.println("Sink executor was terminated.");
+            }
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```

Comment:
generally we should avoid printstacktrace unless absolutely needed. i dont see the need to print stack trace for an interrupted exception. 
wdyt?
also lets clear the interrupt status.